src/D3MHub.sol

```diff
@@ -221,121 +219,118 @@ contract D3MHub {
+        if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) {
+            toUnwind = type(uint256).max; // We make sure to enter the unwind path
         } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        }
+            Line = vat.Line();
+            debt = vat.debt();
+            targetAssets = ilks[ilk].plan.getTargetAssets(currentAssets);
 
-        emit Unwind(ilk, amount, fees);
-    }
+            // Determine if it needs to unwind due to:
+            unchecked {
+                toUnwind = _max(
+                                _max(
+                                    art > lineWad ? art - lineWad : 0, // ilk debt ceiling exceeded
+                                    debt > Line ? _divup(debt - Line, RAY) : 0 // global debt ceiling exceeded
+                                ),
+                                targetAssets < currentAssets ? currentAssets - targetAssets : 0 // plan targetAssets
+                            );
+            }
+        }
 
-    function _fix(bytes32 ilk, address _pool) internal returns (uint256 diff) {
-        (uint256 ink, uint256 art) = vat.urns(ilk, _pool);
-        if (art < ink) {
-            address _vow = vow;
-            diff = ink - art;
-            require(diff <= MAXINT256, "D3MHub/overflow");
-            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
-            vat.grab(ilk, _pool, _pool, _vow, 0, int256(diff));
+        if (toUnwind > 0) {
+            toUnwind = _min(
+                toUnwind,
+                maxWithdraw
+            );
+            require(toUnwind <= MAXINT256, "D3MHub/overflow");
+            _unwind(ilk, _pool, toUnwind);
+        } else {
+            uint256 toWind;
+            // Determine up to which value to wind:
+            // (subtractions are safe as otherwise toUnwind is > 0)
+            unchecked {
+                toWind = _min(
+                            _min(
+                                _min(
+                                    lineWad - art, // amount to reach ilk debt ceiling
+                                    (Line - debt) / RAY  // amount to reach global debt ceiling
+                                ),
+                                targetAssets - currentAssets // plan targetAssets
+                            ),
+                            _pool.maxDeposit() // restricts winding if the pool has a max deposit
+                        );
+            }
+            require(art + toWind <= MAXINT256, "D3MHub/wind-overflow");
```

Comment:
Should document `art <= MAXINT256` as an intended invariant of valid D3M implementations in the README or elsewhere.

Reply:
This was a bit changed. Now the `require` is `art + toWind <= MAXINT256 / RAY`.
Basically the whole logic is now designed now to work with `ink` and `art` being less or equal to that value, so when we have `int256 dtab = dart * rate (RAY)` in `vat.frob` or `vat.grab` it can not overflow.
We also do these explicit `require`s: `ink >= art` and `ink <= MAXINT256 / RAY` at the beginning of the normal path of `exec` and there is a Certora assert in the spec that checks that `ink` can not end up higher than that number after `exec`.